src/ansi-c/c_typecheck_base.cpp

```diff
@@ -739,33 +739,7 @@ void c_typecheck_baset::typecheck_declaration(
             CPROVER_PREFIX "loop_entry is not allowed in preconditions.");
         }
 
-        for(auto &target : code_type.assigns())
-        {
-          typecheck_expr(target);
-
-          if(target.type().id() == ID_empty)
-          {
-            error().source_location = target.source_location();
-            error() << "void-typed targets not permitted" << eom;
-            throw 0;
-          }
-          else if(target.id() == ID_pointer_object)
-          {
-            // skip
-          }
-          else if(!target.get_bool(ID_C_lvalue))
-          {
-            error().source_location = target.source_location();
-            error() << "illegal target in assigns clause" << eom;
-            throw 0;
-          }
-          else if(has_subexpr(target, ID_side_effect))
-          {
-            error().source_location = target.source_location();
-            error() << "assigns clause is not side-effect free" << eom;
-            throw 0;
-          }
-        }
+        typecheck_spec_assigns(code_type.assigns());
```

Comment:
Random place for comment: I would prefer, if at all possible, for tests to be part of this first commit. Those tests should make sure that the newly added parsing and type checking actually works. Also, this newly added support ought to be documented.

Reply:
I've added some description regarding this to the commit message now. Basically the tests (along with the syntax rules for parsing) were introduced in #6196. This PR only adds the parsing _actions_ etc.

Reply:
Is there anyway we can test that this was successful?

Reply:
Yes there are lots of function contracts regression tests that check each branch of the `if/else` structure. They still pass. I just refactored this into a common function that can be used for loop contracts too.